### ISSUE TITLE
Uncaught Error remote.js

### ISSUE THREAD

**Issue by paul321** [electron/electron#7351]
- Electron version: 1.3.5 - 1.4.1
- Operating system: MacOS 

I keep getting this error if i start the app with dev tools open and then hit refresh, from the second time it keeps showing this error and i don't know why happens, first time when i run the app everything is ok. This doesn't happen in 1.2.6

```
remote.js:217 Uncaught Error: Cannot read property 'object' of undefined
TypeError: Cannot read property 'object' of undefined
    at ObjectsRegistry.get (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/objects-registry.js:44:28)
    at EventEmitter.<anonymous> (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:306:31)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:219:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
```

And after the first run the app is full of another error 

```
remote.js:209 Uncaught (in promise) Error: Cannot read property 'object' of undefined
TypeError: Cannot read property 'object' of undefined
    at ObjectsRegistry.get (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/objects-registry.js:44:28)
    at EventEmitter.<anonymous> (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:305:31)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:218:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at ObjectsRegistry.get (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/objects-registry.js:44:28)
    at EventEmitter.<anonymous> (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:305:31)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:218:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at metaToValue (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:209:13)
    at remoteMemberFunction (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:110:18)
    at 
```

**Comment by bengotow** [comment]
Hey! We're also seeing this when running the spec suite for N1. Seems to have just started happening in Electron 1.4.1.

**Comment by MarshallOfSound** [comment]
Can confirm, I've been hitting this as well

**Comment by MarshallOfSound** [comment]
Woops, wrong button :(

**Comment by bengotow** [comment]
Looks like some sort of call to a cleaned-up object through `remote` — I wonder if it'd be a good idea to add the method names to these error messages so it'd be a bit easier to track down: https://github.com/electron/electron/blob/master/lib/browser/rpc-server.js#L306

**Comment by jhgg** [comment]
We're seeing this exact same exception too. Happens when we use ctrl + R to reload the app. 

```
remote.js:176 Uncaught Error: Cannot read property 'object' of undefined
TypeError: Cannot read property 'object' of undefined
    at ObjectsRegistry.get (/Applications/Discord Canary.app/Contents/Resources/electron.asar/browser/objects-registry.js:44:28)
    at EventEmitter.<anonymous> (/Applications/Discord Canary.app/Contents/Resources/electron.asar/browser/rpc-server.js:305:31)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Applications/Discord Canary.app/Contents/Resources/electron.asar/browser/api/web-contents.js:218:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
```

**Comment by kevinsawicki** [comment]
@paul321 @jhgg is this happening consistently or intermittently? Would it be possible to get the app you are seeing this on so I can try and debug it with a custom build? Thanks.

**Comment by jhgg** [comment]
Yeah. Just download Discord https://discordapp.com/download - pop open the inspector - hit ctrl - r a few times and it should happen. Definitely not seeing a 100% repro on it though. Sometimes it happens, others it doesn't. Sometimes hitting Ctrl+R again after the exception happens will cause it to refresh properly.

**Comment by kevinsawicki** [comment]
@jhgg should I use the PTB version? Are you ever seeing it when refreshing the login screen or should I log in?

**Comment by kevinsawicki** [comment]
I can reproduce this by running the `N1` test suite, the failing line is `remote.process.stdout.write(str)` in their spec runner.

This looks to happen intermittently when a `ELECTRON_BROWSER_DEREFERENCE` fires for the `stdout` property while it is being accessed.

So the order I'm seeing is:

- `ELECTRON_BROWSER_MEMBER_GET` fires from renderer to main process for the `stdout` property.
- The main process replies with the metadata information and the id of the object from the cache.
- Before the renderer process receives the reply, a `ELECTRON_BROWSER_DEREFERENCE` fires for the `stdout` property clearing it from the cache.
- When the renderer process receives the metadata and builds up the remote object, it is now using a cache id that is no longer present in the cache so the when `write` is called on the `stdout` remote property, it fails to obtain the `stdout` object from the object cache for the previously returned id.

I can only reproduce this maybe once out of every 20 times tried so it seems to be dependent on a GC run happening at a specific time where the main process is building the metadata for the remote object and the renderer process runs the destructor for it.

**Comment by adamwoodnz** [comment]
We're also getting this on 1.3.3

**Comment by jhgg** [comment]
@kevinsawicki both should work - they are on the same version of Electron now. Based on your findings above - I guess it'd only work when you log in and we start interfacing with remote objects.

**Comment by zcbenz** [comment]
It seems that the reference chain is broken, `write` should reference `stdout`, it probably referenced `process` instead.

**Comment by zcbenz** [comment]
I'm wondering whether we have met a bug of V8, for the `remote.process.stdout.write(str)` call, it works like this underlying:

1. send `GLOBAL` message to get `process`
2. send `MEMBER_GET` message to get `stdout` from `process`
3. send `MEMBER_CALL` message to call `write` from `stdout`

And the bug happens for either step 2) or 3), when `process` is garbage collected before `process.stdout` is called, or when `process.stdout` is garbage collected before `process.stdout.write` is called.

So for the expression `xxx.yyy`, it is possible that the `xxx` can be garbage collected when `xxx.yyy` is being called. This only happens when `yyy` is defined via `Object.defineProperty`.

**Comment by fantasticsid** [comment]
so the workaround for now would be avoid remote calls and only use IPC messages?

**Comment by Eugeny** [comment]
@fantasticsid it seems that if the bug doesn't happen in the first second during webcontents load, it never happens later.

**Comment by paul321** [comment]
Yes when you load the app first time it doesn't happen, only after you
refresh it by using ctrl+r or cmd+r

**Comment by Eugeny** [comment]
In my case, ctrl-r actually "fixes" the issue, i.e. the app reloads without errors.
So in short, I have a ~10% chance of the abovementioned crash when [re]loading the app, but if it happens, reloading it again always helps.

**Comment by paul321** [comment]
Just the opposite case for me

**Comment by KochiyaOcean** [comment]
It sometimes occurs on a webview that run serval hours. All methods of webview (such as refresh, forward, getURL, etc) will return this Error

**Comment by Eugeny** [comment]
@zcbenz not sure if I'm doing it right, but I tried to get and keep a global reference to `process` in the main process and `remote.process` in the renderer process to prevent them from being GCed, but it didn't help.

**Comment by vojtatranta** [comment]
Please please guy, our customers are very angry with this.

Could you not just do something so that in would not throw such errors?

This line: https://github.com/electron/electron/blob/master/lib/browser/objects-registry.js#L44

Is the root of the evil by itself. Yeah I know, there should be object with this id and it should have property `object` but apparantly. There is a bug?

What if we just ifed it out and set logging there so we would know more? Thanks!

@zcbenz thanks!

**Comment by petrbrzek** [comment]
Any news about this one?

**Comment by fantasticsid** [comment]
@vojtatranta ifing things  out will just make things worse in the future

**Comment by paul321** [comment]
Any news?

**Comment by Eugeny** [comment]
@paul it seems to go away when running the production (i.e. built with electron-builder) version of my app. As for the development, I'm just restarting the app when it crashes :(

**Comment by Eugeny** [comment]
@fantasticsid it's still better than the framework being unusable in some cases. Just if it out and spit out a warning instead. This way you can have working software AND fix it for real later.

**Comment by paul321** [comment]
In production you never get this error because the user never CTRL+R or CMD+R so there is no refresh :)

**Comment by petrbrzek** [comment]
I wonder if this (https://github.com/paulcbetts/electron-remote) would help to prevent the issue. @paulcbetts

**Comment by Eugeny** [comment]
@paul321 well, first, I do (own hotkey handler) and, second, the bug does also happen on the first app load sometimes.

**Comment by paul321** [comment]
So we are stuck on 1.2.6 :(

**Comment by albertotagliavini** [comment]
In the piece of code that fails with this error, do you use the api remote.getGlobal(name) to call the remote function or you use the remote.require(module) api??

Because i have seen that changing the implementation, i'm getting no errors.
Before i had a module in global in the main process and i was obtaining it in the renderer process with remote.getGlobal(name) and sometimes i was getting the same error you are talking about.

Now i have changed my code and i'm calling the remote module by remote.require(module).
It seems that i'm getting no errors.

**Comment by paul321** [comment]
I'm using it this way:
var app = require('electron').remote.app;

Can you give us an example from our perspective?

**Comment by yumaogushi** [comment]
@KochiyaOcean I also met the same problem.

**Comment by qmphan** [comment]
I can reproduce this error with the following code:

**main.js:**
"use strict";
let { app, Menu, BrowserWindow } = require('electron');
let mainWindow = null;
global['mytest'] = {'hello': 'world'};
function createWindow() {
    // Create the browser window.
    var path = require('path');
    mainWindow = new BrowserWindow({ width: 800, height: 600,
        webPreferences: {
            'preload': path.resolve(path.join(__dirname, 'preload.js')),
            'nodeIntegration': true
        }
    });
    mainWindow.loadURL(`http://www.google.com`);
    mainWindow.loadURL(`http://www.google.com`); // Calling this line 2nd time causes the error
    
    // Open the DevTools.
    mainWindow.webContents.openDevTools();
    // Emitted when the window is closed.
    mainWindow.on('closed', function () {
        mainWindow = null;
        app.quit();
    });
}
// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow);

**preload.js**
process.once('loaded', () => {
    const remote = require('electron').remote;
    window.jsphone_api = remote.app.jsphone_api;
    console.log('mytest = ', remote.getGlobal('mytest'));
    console.log('preload got called');
})

- When I package it using electron-packager with option --arch=ia32 and run it on a 32bit Windows 7, this error happens every time. You can see in the log the message "Cannot read property 'object' of undefined TypeError: Cannot read property 'object' of undefined...". Note that the error happens because the function mainWindow.loadURL(`http://www.google.com`) is called twice. If I call it only once, there is no error.
- When I run it on a 64 bit Windows 7 machine, there is no error.

**Comment by fantasticsid** [comment]
@Eugeny silently swallowing the exception and expect your app to carry on IMHO is pretty bad.

**Comment by Eugeny** [comment]
@fantasticsid absolutely. That's why I'm saying "spit out a warning instead" in my comment. That let's people carry on using Electron productively until this bug is fixed, without having to resort to patching Electron itself.

**Comment by paul321** [comment]
Any news on this one?
A lot of production apps are crashing because of this bug, it's ridiculous.

**Comment by paul321** [comment]
It's crashing only JS but the app will not work after the error was thrown, we are getting a lot of js errors after the first encounter and the app starts to throw other errors just like this one

**Comment by paul321** [comment]
Thing is in our app we do not transfer any date between processes, the app it's only working on the render process

**Comment by kevinsawicki** [comment]
@paul321 how many calls to the `remote` module is your app making? Are you able to switch them to use `ipcRenderer` until this issue is resolved?

**Comment by paul321** [comment]
Just checked the source code, we have about 20 calls to remote like:
remote,session
remote.getCurrentWindow()
remote.app
...

all of them are prefixed by require('electorn')

**Comment by kevinsawicki** [comment]
> Just checked the source code, we have about 20 calls to remote like:

Do you know which specific ones are throwing this error?

**Comment by paul321** [comment]
nope, because we don't see which call throws the error from render process, we only see the error from main process and no stack trace

**Comment by kevinsawicki** [comment]
> nope, because we don't see which call throws the error from render process, we only see the error from main process and no stack trace

Electron 1.4.8 will be out later today and will contain better error messages for this bug. If you can re-run your app on that version you'll be able to know which remote calls are affected and you can then workaround them by using the `ipcRenderer` module until this issue is completely fixed.

**Comment by paul321** [comment]
Ok thank you, will keep you posted if i can get more info from 1.4.8

**Comment by kevinsawicki** [comment]
> Ok thank you, will keep you posted if i can get more info from 1.4.8

1.4.8 is now available.

**Comment by paul321** [comment]
Here is the error, call is made from this line
App.win = require('electron').remote.getCurrentWindow();

setTimeout(function() {
      App.win.show();
}, 300);

Uncaught Error: Cannot call function 'show' on missing remote object 10
Error: Cannot call function 'show' on missing remote object 10
    at throwRPCError (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\api\web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)

**Comment by paul321** [comment]
Another one

Uncaught (in promise) Error: Cannot call function 'getPath' on missing remote object 18
Error: Cannot call function 'getPath' on missing remote object 18
    at throwRPCError (D:\devel\work\js\microstockr.desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\api\web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at throwRPCError (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous>

**Comment by paul321** [comment]
It seems i cannot call remote functions from render process, even if i put a break point at that line, dev tools console will show me that setProgressBar it's a function

 Uncaught (in promise) Error: Cannot call function 'setProgressBar' on missing remote object 41
Error: Cannot call function 'setProgressBar' on missing remote object 41
    at throwRPCError (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\api\web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at throwRPCError (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:143:17)
    at EventEmitter.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\browser\api\web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at metaToValue (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\renderer\api\remote.js:217:13)
    at remoteMemberFunction (D:\devel\work\js\desktop\node_modules\electron-prebuilt\dist\resources\electron.asar\renderer\api\remote.js:113:18)

**Comment by Qquanwei** [comment]
this problem occur for me , 

my preload.js
```
window.pkg = require('electron').remote.require('./package.json');
```
throw the same error when I access window.pkg

**Comment by paul321** [comment]
I'm using require js in a file like this one and keep a reference to remote objects but after few minutes the objects are not accessible anymore 
```
define([
    'moment',
    'async',
    'src/constants'
], function(moment,async, Constants) {
    return {
        remote: require('electron').remote,
        path: require('path'),
        filesys: require('fs'),
        app: require('electron').remote.app,

        init: function() {
            return this;
        },
     
        thumbs: function() {
            console.log(self.app.getPath('userData'));
        }
});
```

In few minutes then function thumbs is called 'self.app.getPath' with throw an error

**Comment by paul321** [comment]
Even if i make the remote variables scope vars in thumbs function it's still crashing on CTRL + R or CMD+R

**Comment by gewl** [comment]
I've been getting what is (I believe) a related error. I'm using chentsulin's electron-react-boilerplate. Production and dev both work fine, but when I run a test I get this error from electron-json-storage:

```
[dirpath]/node_modules/electron-json-storage/lib/utils.js:30
const app = electron.app || electron.remote.app;
^

TypeError: Cannot read property 'app' of undefined
```

Would love any advice on how to fix this!

**Comment by MarshallOfSound** [comment]
@gewl That's a different error. Looks like either your using tmux or your node_modules install is a bit messed up.

**Comment by gewl** [comment]
Not using tmux! node_modules install may be messed up, but if so, this is the only place that it's creating any problems.

I got it worked out. For future reference, (one possible) answer was to use electron-mocha instead of mocha in the test script. Thanks for the suggestions!

**Comment by etyp** [comment]
@kevinsawicki Please let us know if there's any specific information we can provide to help resolve this issue.

I'm only seeing it on our Windows build now, but previously saw a lot of it on our users with OSX builds:
- Electron versions: 1.4.1 and 1.3.9
- OS: Windows 10

Error + stack:

```
Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying error: Error processing argument at index 0, conversion failure from #<Object> Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying er 
Stack: 
Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying error: Error processing argument at index 0, conversion failure from #<Object>
Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying error: Error processing argument at index 0, conversion failure from #<Object>
    at callFunction (C:\Users\Redacted\AppData\Local\Hive\app-1.2.2\resources\electron.asar\browser\rpc-server.js:226:11)
    at EventEmitter.<anonymous> (C:\Users\Redacted\AppData\Local\Hive\app-1.2.2\resources\electron.asar\browser\rpc-server.js:307:5)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (C:\Users\Redacted\AppData\Local\Hive\app-1.2.2\resources\electron.asar\browser\api\web-contents.js:219:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at metaToValue (C:\Users\Redacted\AppData\Local\Hive\app-1.2.2\resources\electron.asar\renderer\api\remote.js:210:13)
    at remoteMemberFunction (C:\Users\Redacted\AppData\Local\Hive\app-1.2.2\resources\electron.asar\renderer\api\remote.js:111:18)
```

I can also include our preload.js and usage of require if that helps.

**Comment by kevinsawicki** [comment]
@etyp that stack you posted is a different error, not related to this issue.

> Error processing argument at index 0, conversion failure from #<Object>

It looks like `setOverlayIcon` was called with a first argument of something other than a `NativeImage` instance.

**Comment by kevinsawicki** [comment]
Electron 1.4.11 and 1.3.13 are now out with a potential fix for this issue via #8110.

Can people try updating to those versions and see if they can still reproduce this issue? Thanks.

**Comment by Mic75** [comment]
On the project I am working on, we had the exception thrown from remote.js while refreshing the app (we were on 1.4.5 and we had to downgrade to 1.3.3).

After upgrading to 1.4.11, no more exception thrown from remote.js, it looks like the issue is fixed !
Thank you sir !

**Comment by VINTproYKT** [comment]
@kevinsawicki This fixed *my* issue with disappearing remote objects. An interesting coincidence is that I have downloaded 1.4.10 three hours ago and got stuck with this problem. Now I'm all set, thank you! 👍

**Comment by kevinsawicki** [comment]
@paul321 are you able to give the latest releases, 1.3.13 or 1.4.11, a try and see if you still experience this issue? Thanks.

**Comment by paul321** [comment]
I did try it today and it seems fixed :)
Thank you

**Comment by kevinsawicki** [comment]
> I did try it today and it seems fixed :)

Thanks for confirming, will tentatively close this out, it if starts cropping up again we can reopen.

**Comment by paul321** [comment]
I keep getting this error when i reload the application without restart



remote.js:217 Uncaught Error: Cannot call function 'isVisible' on missing remote object 15
Error: Cannot call function 'isVisible' on missing remote object 15
    at throwRPCError (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/Paul/Work/js/desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)

From this piece of code:

`
var win = require('electron').remote.getCurrentWindow();
if(!win.isVisible()) {
    win.show();
}  
`

**Comment by paul321** [comment]
and a lot of random remote errors like this one too:

remote.js:217 Uncaught (in promise) Error: Cannot call function 'getSize' on missing remote object 26
Error: Cannot call function 'getSize' on missing remote object 26
    at throwRPCError (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at throwRPCError (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:339:7)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:231:13)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at metaToValue (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:217:13)
    at remoteMemberFunction (/Users/Paul/Work/js/microstockr.desktop/node_modules/electron-prebuilt/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:113:18)

`
var win = require('electron').remote.getCurrentWindow();
var size = win.getSize();
`

**Comment by paul321** [comment]
Ok my mistake, i didn't updated Electron on this machine :)

**Comment by etyp** [comment]
Updated to Electron 1.4.14 about 2 weeks ago and our users are still seeing this issue @kevinsawicki . See the screenshot below that shows the stack + the Electron version:

![missing-remote-screen](https://cloud.githubusercontent.com/assets/5808998/22354815/15f2d9a2-e3f5-11e6-99fd-e32d1aa0a5f2.png)

Any idea how we can resolve in the short term? Also, is there any chance that something in their user data folder from an earlier, broken version is causing this to happen despite autoupdate moving them to a new version of Electron?

**Comment by kevinsawicki** [comment]
> Any idea how we can resolve in the short term?

Can you include the code running when the error is thrown?

> Also, is there any chance that something in their user data folder from an earlier, broken version is causing this to happen despite autoupdate moving them to a new version of Electron?

No, if the console reports 1.4.14 then the update to that version has occurred.

**Comment by etyp** [comment]
Sorry for the delay @kevinsawicki 

On BrowserWindow instance, we use the `webPreferences.preload` option like so to avoid conflicting `require` statements:

```
// Main Process - main.js

mainWindow = new BrowserWindow({
    webPreferences: {
      nodeIntegration: false,
      preload: path.join(__dirname, 'preload.js')
    }
});
```

Where preload.js is:

```
// preload.js

// Use new namespace for require so that
// other modules can use 'require' properly.
DesktopRequire = require;
delete require;
```


And then on the renderer process (our web app), we define a Desktop object which hold references to select Electron modules like remote:

```
  // /some/fake/path/desktop.js

  const isDesktop = /Electron/.test(navigator.userAgent) && typeof DesktopRequire !== 'undefined';
  export const Desktop = isDesktop ? {} : false;
  if (typeof DesktopRequire === 'undefined') {
    DesktopRequire = function () { return function () { return false; }; };
  }
  // Keep a reference on the window
  // [Attempt to avoid https://github.com/electron/electron/issues/7351]
  window.Desktop = Desktop;
  window.electron = {};

  $(document).ready(() => {
    window.electron = DesktopRequire('electron');
    // ipc for talking to electron
    Desktop.ipc = electron.ipcRenderer;
    // Catch autoupdate errors
    Desktop.ipc.on('autoUpdaterError', (error, message) => {
      console.log(error);
      console.log(message);
    });
    // remote for calling remote methods
    Desktop.remote = electron.remote;
    // Native Image
    Desktop.nativeImage = electron.nativeImage;
    const appVersion = Desktop.remote.app.getVersion();
    const intAppVersion = parseInt(appVersion.replace(/\./g, ''), 10);
    // Dialog file system
    Desktop.dialog = electron.remote.dialog;
    // New browser windows
    Desktop.window = electron.remote.getCurrentWindow();
  });
```

Finally, we have an interval that calls this function with a certain number value to update the badge:

```
// /some/fake/path/set-badge.js
import { Desktop } from '/some/fake/path/desktop.js';
export const setBadge = (text) => {
  const app = window.Desktop.remote.app || Desktop.remote.app;
  const remote = window.Desktop.remote || Desktop.remote;
  const nativeImage = remote.require('electron').nativeImage;

  if (remote.process.platform === 'darwin') {
    app.dock.setBadge('' + text);
  } else if (remote.process.platform === 'win32') {
    const win = remote.getCurrentWindow();

    if (text === '') {
      win.setOverlayIcon(null, '');
      return;
    }

    // Create badge
    const canvas = document.createElement('canvas');
    canvas.height = 140;
    canvas.width = 140;
    const ctx = canvas.getContext('2d');
    ctx.fillStyle = 'red';
    ctx.beginPath();
    ctx.ellipse(70, 70, 70, 70, 0, 0, 2 * Math.PI);
    ctx.fill();
    ctx.textAlign = 'center';
    ctx.fillStyle = 'white';

    if (text.length > 2) {
      ctx.font = '75px sans-serif';
      ctx.fillText('' + text, 70, 98);
    } else if (text.length > 1) {
      ctx.font = '100px sans-serif';
      ctx.fillText('' + text, 70, 105);
    } else {
      ctx.font = '125px sans-serif';
      ctx.fillText('' + text, 70, 112);
    }

    const badgeDataURL = canvas.toDataURL();
    const img = nativeImage.createFromDataURL(badgeDataURL);

    win.setOverlayIcon(img, text);
  }
};
```

We used to not save the object to the `window`, but I added this in an attempt to keep `remote` from being dereferenced. I have a difficult time recreating locally, but we have dozens of users reporting it still happening, plus I see it from time to time.

I'm going to try saving the `Desktop` object to global scope by not declaring `const | let | var` before it, but aside from that I'm unsure what to try.

Any help would be absolutely amazing and much appreciated.

**Comment by paul321** [comment]
We got the same problem when calling the function that is setting the badge icon from code

Uncaught (in promise) Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying error: Error processing argument at index 1, conversion failure from 6
Error: Could not call remote function 'setOverlayIcon'. Check that the function signature is correct. Underlying error: Error processing argument at index 1, conversion failure from 6

If we run the function from dev console like this App.setBadge('1') it's working find, called from other functions doesn't work

**Comment by paul321** [comment]
i think there is a problem setting the badge and i think someone should create a new ticket, check this out:
http://stackoverflow.com/questions/41416052/nativeimage-not-working-in-setoverlayicon-in-electron

**Comment by paul321** [comment]
@etyp i found the error, if you send the text as an integer it's not working, and you will get that error, because in your  win.setOverlayIcon(img, text); the text var could be integer and this will break up and throw that error

**Comment by etyp** [comment]
@paul321 good find, but turns out it's not the issue we're encountering. We convert to string every time, plus we still see the error on Mac AND Windows, despite returning early from that function if on darwin.

Just some of the recent errors thrown:

OSX 10.9.5 sample:
```
Error: Cannot get property 'dock' on missing remote object 4877
```

Windows 7 sample:
```
Error: Cannot get property 'platform' on missing remote object 554
```

Windows 10 sample:
```
Error: Cannot get property 'platform' on missing remote object 554
```

OSX 10.12.13 sample:
```
Error: Cannot get property 'nativeImage' on missing remote object 343
```

OSX 10.12.3 sample:
```
Error: Cannot get property 'dock' on missing remote object 12
```

OSX 10.12.3 sample:
```
Error: Cannot get property 'platform' on missing remote object 76
```

Is anybody else that no longer encounters this using es6 modules?

I know that imports are meant to keep 1 cached version in memory after the first time they're imported, but I'm worried something else is going wrong here (hence the saving references to `window.Whatever`).

**Comment by etyp** [comment]
@kevinsawicki might this have anything to do with deleting the `require` reference on our preload script like this:

```
// preload.js

// Use new namespace for require so that
// other modules can use 'require' properly.
DesktopRequire = require;
delete require;
```

**Comment by whmountains** [comment]
I get this error 100% of the time when running this example project: https://github.com/avocode/electron-windows-autoupdate.

```
Uncaught Error: Could not call remote function 'setFeedURL'. Check that the function signature is correct. Underlying error: Could not get code signature for running application
Error: Could not call remote function 'setFeedURL'. Check that the function signature is correct. Underlying error: Could not get code signature for running application
    at callFunction (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:257:11)
    at EventEmitter.<anonymous> (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:357:5)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:256:37)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at callFunction (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:257:11)
    at EventEmitter.<anonymous> (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:357:5)
    at emitMany (events.js:127:13)
    at EventEmitter.emit (events.js:201:7)
    at WebContents.<anonymous> (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:256:37)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at metaToValue (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:234:13)
    at Object.remoteMemberFunction (/Users/mmc/code/electron-windows-autoupdate/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/renderer/api/remote.js:118:18)
    at file:///Users/mmc/code/electron-windows-autoupdate/index.html:31:15
```

The dependencies in that project are a bit out of date, but after updating everything I still get the same error.  I'm running Mac OS 10.12.4 Sierra, using Node v7.9.0, and installing packages with Yarn 0.22.0.

**Comment by kevinsawicki** [comment]
> Could not get code signature for running application

@whmountains Auto updating on macOS requires the application to be signed, that is why you are seeing this error.

**Comment by whmountains** [comment]
@kevinsawicki Right, so this is a completely different error.  Sorry about that.

I'm not going to delete my comment in case someone else searches that error message and finds this thread like I did.

**Comment by mqtik** [comment]
It's still happening as of  7.1.14
